tee/prototype/blockchain.go

```diff
@@ -5,84 +5,142 @@ package prototype
 
 // PushVerify pushes the block onto the chain, verifying that it is indeed a
-// correct next block.
-//
-// If the chain's current head has the same or a larger block number, a chain
-// reorg is assumed and the current block becomes the new head, discarding all
-// previous blocks.
-func (b *blockchain) PushVerify(block *tee.Block) error {
-	blockNum := block.NumberU64()
-	if len(b.blocks) == 0 {
-		// first block
-		b.offset = blockNum
-		b.blocks = []*tee.Block{block}
-		return nil
+// valid successor block of the previous head. If the block is valid, returns
+// all of the block's deposit and exit events. If the block is invalid, the
+// chain remains unchanged.
+func (b *blockchain) PushVerify(
+	block *tee.Block,
+	params *tee.Parameters,
+	epoch *Epoch,
+) ([]*erdstallDepEvent, []*erdstallExitEvent, error) {
+	if !b.empty() {
+		if err := verifySuccessorBlock(block, b.head); err != nil {
+			return nil, nil, fmt.Errorf("verifying successor block: %v", err)
+		}
 	}
 
-	prev := b.Head()
-	prevNum := prev.NumberU64()
-	if blockNum > prevNum+1 {
-		return fmt.Errorf("intermediate blocks missing, head: %d, block: %d", prevNum, blockNum)
-	} else if blockNum <= prevNum {
-		// reorg
-		prev = b.Block(blockNum - 1)
-		prevNum = prev.NumberU64()
+	deps, exits, err := b.extractEvents(block, params, epoch)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid block events: %w", err)
 	}
 
-	if err := verifyBlock(block, prev); err != nil {
-		return fmt.Errorf("verifying block: %v", err)
+	blockN := block.NumberU64()
+	depEpoch, exitEpoch := params.DepositEpoch(blockN), params.ExitEpoch(blockN)
+
+	if err = verifyDeposits(deps, depEpoch); err != nil {
+		return nil, nil, fmt.Errorf("invalid deposits: %w", err)
+	} else if err = verifyExits(exits, exitEpoch, epoch); err != nil {
+		return nil, nil, fmt.Errorf("invalid exits: %w", err)
 	}
 
-	b.blocks = append(b.blocks[:prevNum+1-b.offset], block)
+	b.head = block
+	return deps, exits, err
+}
+
+// verifySuccessorBlock verifies that next is a valid successor of head.
+func verifySuccessorBlock(next, head *tee.Block) error {
+	nextHash, headHash := next.Header().ParentHash, head.Hash()
+	nextN, headN := next.NumberU64(), head.NumberU64()
+
+	// TODO: Extend validation to test consensus (PoW).
+	//   This function then probably becomes a method on nextchain after we add a
+	//   consensus engine to nextchain.
+	switch true {
+	case nextN != headN+1:
+		return fmt.Errorf("next is not successor, head: %d, next: %d", headN, nextN)
+	case nextHash != headHash:
+		return fmt.Errorf("head header mismatch, expected %x, got: %x", nextHash, headHash)
+	}
 	return nil
 }
 
-// PruneUntil can be used to discard all blocks until the given block number
-func (b *blockchain) PruneUntil(blockNum uint64) {
-	if b.offset >= blockNum {
-		return
+// extractEvents extracts all deposit and exit events from a block.
+func (b *blockchain) extractEvents(
```

Comment:
removed epoch, but kept the functions as freestanding for now. I don't want to mess around in the rest of the code base outside the tee/prototype directory, as this would also make me rework of the whole tee.Block API. I'll open an issue for this.